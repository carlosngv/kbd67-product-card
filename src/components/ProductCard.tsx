import React, { createContext } from "react";

import { InitialValues, onChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/products.interfaces';

import { useProduct } from '../hooks/useProduct';
import styles from '../styles/styles.module.css';


export const ProductContext = createContext( {} as ProductContextProps );
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[];
    children?: (args: ProductCardHandlers ) => React.JSX.Element,
    className?: string;
    style?: React.CSSProperties;
    onChange?: ( args: onChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}


export const ProductCard = ( { product, children, className, style, onChange, value, initialValues }: Props ) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct( { onChange, product, value, initialValues } );

    return (
        <Provider value={ { counter, increaseBy, product, value, maxCount } }>
            <div style={ style } className={ `${ styles.productCard } ${ className }` }>
                { children!({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    increaseBy,
                    product,
                    reset
                }) }
                
            </div>
        </Provider>
    )
}
