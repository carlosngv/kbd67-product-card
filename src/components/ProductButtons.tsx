
import React from 'react';

import styles from '../styles/styles.module.css';

import { useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";

export interface Props {
    className?: string;
    activeButtonClass?: string;
    style?: React.CSSProperties;

}

export const ProductButtons = ({ className, style }: Props ) => {

    // TODO: desestructurar maxCount
    const { counter, increaseBy, maxCount } = useContext(ProductContext);
    
    // TODO: useCallback isMaxReached cuando el counter cambie, retorna un boolean
    const isMaxReached = useCallback(
        () => !!maxCount && maxCount === counter
        , [ counter ]);

    return (
        <div style={ style } className={ `${ styles.buttonsContainer } ${ className }` }>
            <button
                className={ styles.buttonMinus}
                onClick={ () => increaseBy( -1 ) }
            >-</button>
            <div
                className={ styles.countLabel}
            >{ counter }</div>
            <button
                className={ `${ styles.buttonAdd } ${ isMaxReached() && styles.disabled }` }
                onClick={ () => increaseBy( 1 ) }
                disabled={ isMaxReached() }
            >+</button>
        </div>
    );
}
