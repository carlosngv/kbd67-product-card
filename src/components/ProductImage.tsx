
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg'

import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";

export interface Props {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductImage = ( { img = noImage, className, style }: Props ) => {

    const { product } = useContext(ProductContext);

    return (
        <img style={ style } className={ ` ${ styles.productImg } ${ className }` } src={ product.img ? product.img : img } alt="Product image"/>
    );

}
