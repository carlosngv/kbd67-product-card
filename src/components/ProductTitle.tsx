import styles from '../styles/styles.module.css';

import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";


export interface Props { 
    title?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductTitle = ( { title, className, style } : Props ) => {

    const { product } = useContext(ProductContext);
    let productTitle: string;

    if( title ) {
        productTitle = title;
    } else if( product.title ){
        productTitle = product.title
    } else {
        productTitle = 'Product';
    }


    return (
        <span style={ style } className={ `${ styles.productDescription } ${ className }`  }>{ productTitle }</span>
    )
}
