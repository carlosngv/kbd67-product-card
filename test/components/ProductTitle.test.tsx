import * as React from 'react';
import renderer from "react-test-renderer";

// import {render, screen } from '@testing-library/react'

import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe( 'ProductTitle', () => {
    test('should show component with custom title', () => {
        const wrapper = renderer.create(
            <ProductTitle title="Custom Product" /> 
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
    
    test('should show component with product name', async () => {
        const wrapper = renderer.create(
            <ProductCard product={ product1 }>
                {
                    () => (
                        <ProductTitle />
                    )
                }
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
        
    });

});
