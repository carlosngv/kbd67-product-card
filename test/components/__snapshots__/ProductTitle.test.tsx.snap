// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should show component with custom title 1`] = `null`;

exports[`ProductTitle should show component with product name 1`] = `null`;
